builder/main.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mitchellh/cli"
+	"github.com/spf13/viper"
+)
+
+func main() {
+	c := &cli.CLI{
+		Name:       "syntax",
+		Version:    "",
+		Args:       os.Args[1:],
+		HelpWriter: os.Stdout,
+	}
+
+	ui := &cli.ColoredUi{
+		ErrorColor: cli.UiColorRed,
+		WarnColor:  cli.UiColorYellow,
+		Ui: &cli.BasicUi{
+			Writer:      os.Stdout,
+			Reader:      os.Stdin,
+			ErrorWriter: os.Stderr,
+		},
+	}
+
+	c.Commands = map[string]cli.CommandFactory{
+		"build": func() (cli.Command, error) {
+			return &BuildCommand{
+				Ui: ui,
+			}, nil
+		},
+	}
+
+	exitStatus, err := c.Run()
+	if err != nil {
+		ui.Error("Error: " + err.Error())
+	}
+
+	os.Exit(exitStatus)
+}
+
+type BuildCommand struct {
+	Ui cli.Ui
+}
+
+func (c *BuildCommand) Help() string {
+	helpText := `Usage: syntax build
+
+` + c.Synopsis()
+	return strings.TrimSpace(helpText)
+}
+
+func (c *BuildCommand) Synopsis() string {
+	return "Builds syntax files"
+}
+
+func (c *BuildCommand) Run(args []string) int {
+	wd, err := os.Getwd()
+	if err != nil {
+		return 1
+	}
+
+	// The _main.yml file is where all rules in the Patterns and the Repository are defined
+	// Each product file provides specific overrides for rules defined in _main.yml
+	// Here we build a main Viper instance to store the config we will merge in product specific maps to
+	mainFile := filepath.Join(wd, "../src/_main.yml")
+	fmt.Printf("Pulling in main template: %s\n", mainFile)
+	mainViper := viper.New()
+	mainViper.SetConfigFile(mainFile)
+	if err := mainViper.ReadInConfig(); err == nil {
+		fmt.Printf("Merging config file: %v\n", mainViper.ConfigFileUsed())
+	} else {
+		fmt.Printf("Error reading config: %v\n", err)
+		return 1
+	}
+
+	// For each product defined, read the yml and merge into the main Viper instance
+	products := []string{"hcl", "terraform"}
+	for _, product := range products {
+		fmt.Printf("Evaluating %s\n", product)
+		productFile := filepath.Join(wd, fmt.Sprintf("../src/%s.yml", product))
+		fmt.Printf("Processing: %s\n", productFile)
+
+		// Create a product Viper instance that reads each product yml for rules
+		productV := viper.New()
+		productV.SetConfigFile(productFile)
+		if err := productV.ReadInConfig(); err == nil {
+			fmt.Printf("Merging config file: %v\n", productV.ConfigFileUsed())
+		} else {
+			fmt.Printf("Error reading config: %v\n", err)
```

Comment:
That depends on whether you want to stop building all files because one has an error, or do you want to continue building all files and print the error for the single file that has a problem.

I chose the pattern of continuing to build files and display errors for each file that fails.

Reply:
My bad for not being too specific here - sorry.

I was more referring to the fact that we end up printing the error to stdout (as opposed to stderr) - which now seems fixed through the use of the Ui package 👍🏻  - and end up returning `0` (success) exit code. Continuing  sounds like a good approach - but I think will require collection of the errors via go-multierror as mentioned and checking `result.ErrorOrNil()` to decide whether we return `0` or `1` exit code.

https://github.com/hashicorp/go-multierror#usage